### ISSUE TITLE
Support chainer's LSTM

### ISSUE THREAD

**Issue by Kiikurage** [mil-tokyo/webdnn#365]
# Support chainer's LSTM

## Assumption, Limitation.

1. Length of series is fixed.
2. `chainer.functions.LSTM` is called through `chainer.links.LSTM`, not directly.
3. Cell state are initialized before when the link is called.
	
	- Sometimes it's need to keep the state. For example, in text generation, graph are executed iteratively with keeping the state. In this case, we have to use `chainer.functions.StatelessLSTM` and keep the state explicitly.

## Procedure:

```
# Chainer computation graph of LSTM Link
	
       x -+ 
          |                               c -+          +-> c_new
 w_input -+--[Linear]---> wx -+              |          |
          |  "linear1"        |              +--[LSTM]--+
       b -+                   |              |   "fn"   |
                              +--[Add]--> g -+          +-> h_new
       h -+                   |  "add"
          |                   |
          +--[Linear]---> wh -+
          |  "linear2"
w_hidden -+
```

### Convert chainer operator into *unrolled* form.

chainer's `LSTM` functino can be converted into WebDNN `LSTM` operator in the follow steps. Let the chainer lstm link be `link`, chainer lstm function be `fn`, and the WebDNN lstm operator be `op`:

`fn` must have 2 inputs variables, `g` and `c`.

1. Get indirect input variables required for `op`: `x`, `w_input`, `w_hidden`, `b`, and `h` 

	1. If `isinstance(g.creator, Linear)`, it means that it's first time step. In this case, `op` is called with `initial_h=None`. However, in this case `w_hidden` is unknown. Therefore handler passes a dummy variable to `op`, and append `ChainerLSTMFirstTime` attribute into `op`.

	2. If `isinstance(g.creator, Add)`, All variable can be retrieved.

	3. Otherwise, `fn` is not called through `link`. Raise ValueError.

2. Check `assert x.order == OrderNC` and reshape `x` into `shape=(N, 1, C), order=OrderNTC`.

3. Convert direct input variables: `c`

	1. If `c.creator is None`, it means that it's first time step.

		1. Check `isinstance(g.creator, Linear) or (h.creator is None)`. Otherwise, it means that cell state is modified from outside of `link`. raise ValueError.

		2. If `np.all(c.data == 0)`, `op` is called with `initial_c = None`.

		3. Otherwise, `c` must be converted into WebDNN Variable `v_c`, and `op` is called with `initial_c = v_c`.

	2. Otherwise, this cell state is generated at previous time step. It must be already converted.

4. Create and call WebDNN `LSTM` operator with `return_sequences=False, activation="tanh", recurrent_activation="sigmoid"`.
	
### Roll up operators into single LSTM operator

Above procedure convert LSTM as *unrolled* form. By rolling up into single LSTM operators, it makes more efficient.

The optimizer rule `RollUpChainerLSTM` is consisted by follow steps:

	//TODO

**Comment by milhidaka** [comment]
WebDNN has been completely rewritten as **v2.0.0-alpha**, a browser-based ONNX inference runtime (load an ONNX model directly in the browser; no Python conversion step). The v1 Keras / Chainer / TensorFlow / Caffe converters and the old backends that this issue refers to no longer exist, so this report is no longer actionable against the current codebase.

Closing as *not planned*. Thanks for the report!
